### ISSUE TITLE
`brew tap-pin` doesn't work, always installs from core.

### ISSUE THREAD

**Issue by philwo** [Homebrew/brew#5218]
**Please note we will close your issue without comment if you delete, do not read or do not fill out the issue checklist below and provide ALL the requested information. If you repeatedly fail to use the issue template, we will block you from ever submitting issues to Homebrew again.**

- [x] are reporting a bug others will be able to reproduce and not asking a question. If you're not sure or want to ask a question do so on our Discourse: https://discourse.brew.sh
- [x] ran a `brew` command and reproduced the problem with multiple formulae? If it's a problem with a single, official formula (not cask) please file this issue at Homebrew/homebrew-core: https://github.com/Homebrew/homebrew-core/issues/new/choose. If it's a `brew cask` problem please file this issue at https://github.com/Homebrew/homebrew-cask/issues/new/choose. If it's a tap (e.g. Homebrew/homebrew-php) problem please file this issue at the tap.
- [x] ran `brew update` and can still reproduce the problem?
- [x] ran `brew doctor`, fixed all issues and can still reproduce the problem?
- [x] ran `brew config` and `brew doctor` and included their output with your issue?

To help us debug your issue please explain:
- What you were trying to do (and why)

I want to `brew tap bazelbuild/tap`, then `brew tap-pin bazelbuild/tap` that repo, then `brew install bazel` and have `bazel` come from the pinned tap.

- What happened (include command output)

Brew ignored the pin and installed Bazel from the core repo:

```
philwo@philwo-imac ~ brew update
Updated 2 taps (homebrew/core and homebrew/cask).
...

philwo@philwo-imac ~ brew tap bazelbuild/tap
==> Tapping bazelbuild/tap
Cloning into '/usr/local/Homebrew/Library/Taps/bazelbuild/homebrew-tap'...
remote: Enumerating objects: 9, done.
remote: Counting objects: 100% (9/9), done.
remote: Compressing objects: 100% (9/9), done.
remote: Total 9 (delta 0), reused 3 (delta 0), pack-reused 0
Unpacking objects: 100% (9/9), done.
Tapped 2 formulae (38 files, 47.4KB).

philwo@philwo-imac ~ brew tap-pin bazelbuild/tap
==> Pinned bazelbuild/tap

# Note the weird message about git stash (I haven't modified /usr/local/Homebrew,
# nor have stashed changes there. See below how to repro this.)
philwo@philwo-imac ~ brew install bazel
To restore the stashed changes to /usr/local/Homebrew run:
  'cd /usr/local/Homebrew && git stash pop'
==> Downloading https://homebrew.bintray.com/bottles/bazel-0.18.0.mojave.bottle.tar.gz
==> Downloading from https://akamai.bintray.com/73/7360b538438b92a875b46470b3f0a677917252e3b4e070b5cc7a0d6a7a05ff17?__gda__=exp=
######################################################################## 100.0%
==> Pouring bazel-0.18.0.mojave.bottle.tar.gz
==> Caveats
Bash completion has been installed to:
  /usr/local/etc/bash_completion.d

zsh completions have been installed to:
  /usr/local/share/zsh/site-functions
==> Summary
🍺  /usr/local/Cellar/bazel/0.18.0: 12 files, 117.3MB

(Wait, that's the wrong Bazel.)

philwo@philwo-imac ~ brew remove bazel
Uninstalling /usr/local/Cellar/bazel/0.18.0... (12 files, 117.3MB)

philwo@philwo-imac ~ cd /usr/local/Homebrew/
philwo@philwo-imac /usr/local/Homebrew (stable) git status
On branch stable
nothing to commit, working tree clean

philwo@philwo-imac /usr/local/Homebrew (stable) git stash list
stash@{0}: WIP on stable: fdd8e9b7c Merge pull request #5148 from MikeMcQuaid/travis-only-macos

philwo@philwo-imac /usr/local/Homebrew (stable) git stash show stash@{0} 
(Nothing.)

philwo@philwo-imac /usr/local/Homebrew (stable) git stash clear

# OK, something with the pin must be wrong. Let's try to find out what's going on.
philwo@philwo-imac /usr/local/Homebrew (stable) brew tap --list-pinned 
(Nothing.)

philwo@philwo-imac /usr/local/Homebrew (stable) brew tap-pin bazelbuild/tap
==> Pinned bazelbuild/tap

# There's that message again! And the pin is still missing.
philwo@philwo-imac /usr/local/Homebrew (stable) brew tap --list-pinned 
To restore the stashed changes to /usr/local/Homebrew run:
  'cd /usr/local/Homebrew && git stash pop'

# No idea what's going on.
```

- What you expected to happen

After running `brew update`, `brew tap`, `brew tap-pin` and `brew install`, Brew would install Bazel from the pinned repo and not from Core.

- Step-by-step reproduction instructions (by running `brew` commands)

See above. Also, here's `brew config` and `brew doctor` (run after all the above commands):

```
philwo@philwo-imac ~ brew doctor
Your system is ready to brew.

philwo@philwo-imac ~ brew config
HOMEBREW_VERSION: 1.8.0
ORIGIN: https://github.com/Homebrew/brew
HEAD: fdd8e9b7cf1e53322492e1e90f685c19bdfde506
Last commit: 9 days ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: febbf812cc2e65a19d1aea835121d6dd3631578a
Core tap last commit: 3 hours ago
HOMEBREW_PREFIX: /usr/local
CPU: octa-core 64-bit skylake
Homebrew Ruby: 2.3.7 => /System/Library/Frameworks/Ruby.framework/Versions/2.3/usr/bin/ruby
Clang: 10.0 build 1000
Git: 2.19.1 => /usr/local/bin/git
Curl: 7.54.0 => /usr/bin/curl
Java: 1.8.0_181-zulu-8.31.0.1
macOS: 10.14-x86_64
CLT: N/A
Xcode: 10.1
```

Note that we have machines where this is 100% reproducible and other machines where this is never reproducible. We also have user reports from Bazel users who experienced this on their machines, and others who couldn't repro it.

I believe this is much more likely to happen on a machine where Brew has been installed a long time ago. I can't remember repro'ing it on a freshly installed Mac where I just installed Brew.

**Comment by MikeMcQuaid** [comment]
I cannot reproduce this. Please try to figure out where it's reproducible and where it is not. I'd suggest looking at the `brew config`, `brew list --versions` and `brew doctor` differences between people affected and those unaffected.

> # Note the weird message about git stash (I haven't modified /usr/local/Homebrew, # nor have stashed changes there. See below how to repro this.)

I cannot reproduce this either.

**Comment by philwo** [comment]
Here's my script how to reproduce this on my machine. However, it just stopped working and I noticed that you released 1.8.1. After my usual Homebrew installation updated to 1.8.1, I no longer can repro the error there, too, so I guess it was fixed by a commit.

```
#!/bin/bash

set -eux

cd "$HOME"
rm -rf homebrew
git clone https://github.com/Homebrew/brew homebrew
cd homebrew
git checkout -b stable 1.8.0
bin/brew tap bazelbuild/tap
bin/brew tap-pin bazelbuild/tap
bin/brew tap --list-pinned
bin/brew config
```

Let's just close this and if I ever see it again, I'll ping this issue?

**Comment by MikeMcQuaid** [comment]
👍
